### PR TITLE
IGNITE-20722 Support skipCopies for DumpReader

### PR DESCRIPTION
**Pull request by yurinaryshkin**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java [hunk 1]
```diff
@@ -89,13 +90,13 @@ public DumpReader(DumpReaderConfiguration cfg, IgniteLogger log) {
 
                 Map<Integer, List<String>> grpToNodes = new HashMap<>();
 
-                Set<Integer> cacheGroupIds = cfg.cacheGroupNames() != null 
+                Set<Integer> cacheGrpIds = cfg.cacheGroupNames() != null
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java:93]: Seems these changes not related to the ticket. Please, revert.

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java [hunk 2]
```diff
@@ -108,11 +109,22 @@ public DumpReader(DumpReaderConfiguration cfg, IgniteLogger log) {
...
 
+                Map<Integer, Set<Integer>> groups = cfg.skipCopies() ? new HashMap<>() : null;
+
                 for (Map.Entry<Integer, List<String>> e : grpToNodes.entrySet()) {
                     int grp = e.getKey();
 
                     for (String node : e.getValue()) {
                         for (int part : dump.partitions(node, grp)) {
+                            if (groups != null && !groups.computeIfAbsent(grp, key -> new HashSet<>()).add(part)) {
+                                if (log.isDebugEnabled()) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java:120]: Let's log on info level.

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReaderConfiguration.java [hunk 3]
```diff
@@ -58,12 +58,15 @@ public class DumpReaderConfiguration {
     /** Cache group names. */
     private String[] cacheGroupNames;
 
+    /** Skip copies. */
+    private boolean skipCopies;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/dump/DumpReaderConfiguration.java:62]: can be final?

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java [hunk 4]
```diff
@@ -108,11 +109,22 @@ public DumpReader(DumpReaderConfiguration cfg, IgniteLogger log) {
 
                 AtomicBoolean skip = new AtomicBoolean(false);
 
+                Map<Integer, Set<Integer>> groups = cfg.skipCopies() ? new HashMap<>() : null;
+
                 for (Map.Entry<Integer, List<String>> e : grpToNodes.entrySet()) {
                     int grp = e.getKey();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java:115]: YOu can init group set here and omit check on each part:

```
groups.computeIfAbsent(grp, key -> new HashSet<>())
```
```
                            if (groups != null && !groups.get(grp).add(part)) {

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java (hunk 1 above)
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java]: Reverted

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java (hunk 2 above)
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java]: Changed

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReaderConfiguration.java (hunk 3 above)
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/dump/DumpReaderConfiguration.java]: Yes. Updated.

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java (hunk 4 above)
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java:118]: Changed.

**Review by nizhikov — APPROVED** [review]

**Pull request merged**